src/wp-includes/link-template.php

```diff
@@ -1895,109 +1920,131 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
+		 * Filters the JOIN clause in the SQL for an adjacent post query.
+		 *
+		 * The dynamic portion of the hook name, `$adjacent`, refers to the type
+		 * of adjacency, 'next' or 'previous'.
+		 *
+		 * Possible hook names include:
+		 *
+		 *  - `get_next_post_join`
+		 *  - `get_previous_post_join`
+		 *
+		 * @since 2.5.0
+		 * @since 4.4.0 Added the `$taxonomy` and `$post` parameters.
+		 *
+		 * @param string       $join           The JOIN clause in the SQL.
+		 * @param bool         $in_same_term   Whether post should be in the same taxonomy term.
+		 * @param int[]|string $excluded_terms Array of excluded term IDs. Empty string if none were provided.
+		 * @param string       $taxonomy       Taxonomy. Used to identify the term used when `$in_same_term` is true.
+		 * @param WP_Post      $post           WP_Post object.
+		 */
+		$join = apply_filters( "get_{$adjacent}_post_join", $join, $in_same_term, $excluded_terms, $taxonomy, $post );
 
-	wp_cache_set( $query_key, $result, 'counts' );
+		/**
+		 * Filters the WHERE clause in the SQL for an adjacent post query.
+		 *
+		 * The dynamic portion of the hook name, `$adjacent`, refers to the type
+		 * of adjacency, 'next' or 'previous'.
+		 *
+		 * Possible hook names include:
+		 *
+		 *  - `get_next_post_where`
+		 *  - `get_previous_post_where`
+		 *
+		 * @since 2.5.0
+		 * @since 4.4.0 Added the `$taxonomy` and `$post` parameters.
+		 *
+		 * @param string       $where          The `WHERE` clause in the SQL.
+		 * @param bool         $in_same_term   Whether post should be in the same taxonomy term.
+		 * @param int[]|string $excluded_terms Array of excluded term IDs. Empty string if none were provided.
+		 * @param string       $taxonomy       Taxonomy. Used to identify the term used when `$in_same_term` is true.
+		 * @param WP_Post      $post           WP_Post object.
+		 */
+		$where = apply_filters( "get_{$adjacent}_post_where", $where, $in_same_term, $excluded_terms, $taxonomy, $post );
 
-	if ( $result ) {
-		$result = get_post( $result );
+		/**
+		 * Filters the ORDER BY clause in the SQL for an adjacent post query.
+		 *
+		 * The dynamic portion of the hook name, `$adjacent`, refers to the type
+		 * of adjacency, 'next' or 'previous'.
+		 *
+		 * Possible hook names include:
+		 *
+		 *  - `get_next_post_sort`
+		 *  - `get_previous_post_sort`
+		 *
+		 * @since 2.5.0
+		 * @since 4.4.0 Added the `$post` parameter.
+		 * @since 4.9.0 Added the `$order` parameter.
+		 *
+		 * @param string $order_by The `ORDER BY` clause in the SQL.
+		 * @param WP_Post $post    WP_Post object.
+		 * @param string  $order   Sort order. 'DESC' for previous post, 'ASC' for next.
+		 */
+		$sort = apply_filters( "get_{$adjacent}_post_sort", $orderby . ' ' . $limits, $post, $order );
+
+		$old_request = "
+			SELECT $found_rows $distinct $fields
+			FROM {$wpdb->posts} $join
+			$where
+			$groupby
+			$sort
+		";
+
+		return $old_request;
+	};
+
+	add_filter( 'posts_request', $callback, 1, 3 );
```

Comment:
Can this use `post_clauses` so `posts_request` doesn't need to be modified?

Reply:
I am not sure if we can. 

Conder this line https://github.com/WordPress/wordpress-develop/blob/291bf03ecf4a19b31463858607d0985c6b65363d/src/wp-includes/link-template.php#L1977

This is a mixture of limit and order. 

https://github.com/WordPress/wordpress-develop/blob/291bf03ecf4a19b31463858607d0985c6b65363d/src/wp-includes/class-wp-query.php#L2906

https://github.com/WordPress/wordpress-develop/blob/291bf03ecf4a19b31463858607d0985c6b65363d/src/wp-includes/class-wp-query.php#L2909

I am not sure if we can maintain the sort filter. We would have to create a string that mixes order and limit that break the out again. 

This would break filters like [this](https://plugins.trac.wordpress.org/browser/post-types-order/trunk/include/class.functions.php#L233)  or [this](https://github.com/hijiriworld/intuitive-custom-post-order/blob/cab853fec17e3b4dc7362f6778dfcc6f332ef281/intuitive-custom-post-order.php#L649-L660).

Reply:
Thanks for the explanation.

Are you able to add some tests for filters like those and compare the runs against trunk and this branch to make sure they work as expected?

It's possible the `AS p` in the original code will cause problems and using low level caching is more complicated than I first thought.

Reply:
The thing is where, we ARE changing the query that run in this filter. Weather it be the `AS p`, `NNER JOIN $wpdb->term_relationships AS tr ON p.ID `. If you are using these filters, it may also mean that cache invalidation may not work as expected. If you join to another table for example that does not invalidate in the same way. I have seen this done in multilingle setups. Maybe the answer is to detect if these filters are in use.

Reply:
Another solution, might to detect if these filters are in use then WP_Query. I have put together a PR to show this idea here. https://github.com/WordPress/wordpress-develop/pull/3743